File: prob1.java
// Time Complexity : O(N^2)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes


class Solution {
    public String customSortString(String order, String s) {
        int freq[] = new int [26];
        String result = "";
        for (int i = 0; i < s.length(); i++) {
            freq[s.charAt(i) - 'a']++;
        }
        for (int i = 0; i < order.length(); i++) {
            while (freq[order.charAt(i) - 'a'] > 0) {
                result+= order.charAt(i);
                freq[order.charAt(i) - 'a']--;
            }
        }
        for (int i = 0;i < 26; i++) {
            while (freq[i] > 0) {
                result+= (char)(i + 'a');
                freq[i]--;
            }
        }
        return result;
    }
}

File: prob1.py
# Time Complexity :O(n)
# Space Complexity :O(1)
# Leet Code: Yes

class Solution:
    def customSortString(self, order: str, s: str) -> str:
        hashmap = {}
        res = ''

        for char in s:
            if char not in hashmap:
                hashmap[char] = 1
            else:
                hashmap[char] += 1

        for char in order:
            if char in hashmap:
                while hashmap[char] != 0:
                    res += char
                    hashmap[char] -= 1
                hashmap.pop(char)
        
        for char in hashmap:
            while hashmap[char] != 0:
                res += char
                hashmap[char] -= 1

        return res
